tests/test_conn.py

```diff
@@ -247,6 +247,34 @@ async def test_close_disconnects_connection(self):
         conn.close()
         self.assertFalse(conn.connected())
 
+    @run_until_complete
+    async def test_semi_broken_connection(self):
+        """This testcase tries to replicate what happens on connection to semi-broken AWS MSK broker.
+
+        Connection could be established, but then some timeout occurs, and connection is not being destroyed
+        properly, producing 'Unclosed AIOKafkaConnection' and calling loop exception handler, which could lead to
+        failing the whole app for example if we're running via aiorun with `stop_on_unhandled_errors=True`.
+        """
+        host, port = self.kafka_host, self.kafka_port
+
+        def mock_connect(slf):
+            slf._reader = mock.MagicMock()
+            slf._reader.at_eof = mock.MagicMock(return_value=False)
+            raise asyncio.TimeoutError
+
+        # We cannot catch ResourceWarning from __del__, so just monitor `close()` call
+        close_called = False
+
+        def mock_close(slf):
+            slf._reader = None
+            nonlocal close_called
+            close_called = True
+
+        with mock.patch.object(AIOKafkaConnection, 'connect', mock_connect):
+            with mock.patch.object(AIOKafkaConnection, 'close', mock_close):
```

Comment:
Is it possible to reproduce the problem without mocking?

Reply:
I don't know how exactly kafka broker can be brought up to exactly this broken state unfortunately. We've got this situation multiple times on Amazon MSK when kafka cluster was under high load, which caused failures on some brokers + "under replicated partitions" issues.

Reply:
Without it we can be never sure, that we actually fixed the problem. Any script that eventually fails is fine, not necessary suitable for unit tests one.

Reply:
If you remove exception handling from `create_conn` and run `python -m pytest -v --log-cli-level=INFO -s --docker-image aiolibs/kafka:2.12_2.2.2 tests/test_conn.py::ConnIntegrationTest::test_semi_broken_connection` there will be original error logged in live log as:

```
Exception ignored in: <function AIOKafkaConnection.__del__ at 0x7f01a90fc3b0>
Traceback (most recent call last):
  File "/home/vvk/devel/aiokafka/aiokafka/conn.py", line 194, in __del__
    source=self)
ResourceWarning: Unclosed AIOKafkaConnection <AIOKafkaConnection host=127.0.0.1 port=43803>
```

So the unittest is still mimics the issue. Basically any exception from `connect()` after `self._read_task` is set is a problem.

Reply:
Are there other possible race conditions here? Like initializing `self._reader`, but not `self._read_task`?

Reply:
No I think not. The most possible source of exceptions is later call `await self._do_version_lookup()`